File: cpp/src/event_log.cpp
#include "lumina_lob/event_log.hpp"

#include <optional>
#include <string>
#include <unordered_map>

namespace lumina_lob {

int64_t EventLog::next_id() {
    return ++counter_;
}

int64_t EventLog::now_ns() {
    auto now = std::chrono::steady_clock::now();
    return std::chrono::duration_cast<std::chrono::nanoseconds>(now.time_since_epoch()).count();
}

Event EventLog::log_add(int64_t order_id,
                        const std::string& side,
                        std::optional<int64_t> price,
                        int64_t qty,
                        std::optional<int64_t> best_bid,
                        std::optional<int64_t> best_ask) {
    Event ev{
        next_id(),
        now_ns(),
        EventType::ADD,
        order_id,
        side,
        price,
        qty,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        best_bid,
        best_ask,
    };
    events_.push_back(ev);
    return ev;
}

Event EventLog::log_cancel(int64_t order_id,
                           std::optional<int64_t> best_bid,
                           std::optional<int64_t> best_ask) {
    Event ev{
        next_id(),
        now_ns(),
        EventType::CANCEL,
        order_id,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        best_bid,
        best_ask,
    };
    events_.push_back(ev);
    return ev;
}

Event EventLog::log_modify(int64_t order_id,
                           int64_t new_qty,
                           std::optional<int64_t> best_bid,
                           std::optional<int64_t> best_ask) {
    Event ev{
        next_id(),
        now_ns(),
        EventType::MODIFY,
        order_id,
        std::nullopt,
        std::nullopt,
        new_qty,
        std::nullopt,
        std::nullopt,
        std::nullopt,
        best_bid,
        best_ask,
    };
    events_.push_back(ev);
    return ev;
}

Event EventLog::log_fill(int64_t order_id,
                         int64_t counterparty_id,
                         int64_t trade_qty,
                         int64_t price,
                         const std::string& side,
                         int64_t filled_qty,
                         std::optional<int64_t> best_bid,
                         std::optional<int64_t> best_ask) {
    Event ev{
        next_id(),
        now_ns(),
        EventType::FILL,
        order_id,
        side,
        price,
        std::nullopt,
        filled_qty,
        counterparty_id,
        trade_qty,
        best_bid,
        best_ask,
    };
    events_.push_back(ev);
    return ev;
}

std::vector<std::unordered_map<std::string, std::string>> EventLog::to_dicts() const {
    std::vector<std::unordered_map<std::string, std::string>> result;
    result.reserve(events_.size());
    for (const auto& e : events_) {
        std::unordered_map<std::string, std::string> d;
        d["event_id"] = std::to_string(e.event_id);
        d["timestamp_ns"] = std::to_string(e.timestamp_ns);
        d["event_type"] = to_string(e.event_type);
        d["order_id"] = std::to_string(e.order_id);
        d["side"] = e.side.value_or("");
        d["price"] = e.price ? std::to_string(*e.price) : "";
        d["qty"] = e.qty ? std::to_string(*e.qty) : "";
        d["filled_qty"] = e.filled_qty ? std::to_string(*e.filled_qty) : "";
        d["counterparty_id"] = e.counterparty_id ? std::to_string(*e.counterparty_id) : "";
        d["trade_qty"] = e.trade_qty ? std::to_string(*e.trade_qty) : "";
        d["best_bid"] = e.best_bid ? std::to_string(*e.best_bid) : "";
        d["best_ask"] = e.best_ask ? std::to_string(*e.best_ask) : "";
        result.push_back(std::move(d));
    }
    return result;
}

} // namespace lumina_lob

File: lumina_lob/core/event_log.py
"""Nanosecond-precision event journal for order book lifecycle."""
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto
from time import perf_counter_ns


class EventType(Enum):
    ADD = auto()
    CANCEL = auto()
    MODIFY = auto()
    FILL = auto()


@dataclass
class Event:
    """Single order book event."""

    event_id: int
    timestamp_ns: int
    event_type: EventType
    order_id: int
    side: str | None = None
    price: float | None = None
    qty: int | None = None
    filled_qty: int | None = None
    counterparty_id: int | None = None
    trade_qty: int | None = None
    best_bid: float | None = None
    best_ask: float | None = None


class EventLog:
    """Append-only journal of all book events."""

    def __init__(self) -> None:
        self.events: list[Event] = []
        self._counter: int = 0

    def _next_id(self) -> int:
        self._counter += 1
        return self._counter

    def _now_ns(self) -> int:
        return perf_counter_ns()

    def log_add(self, order_id: int, side: str, price: float | None, qty: int, best_bid: float | None, best_ask: float | None) -> Event:
        ev = Event(
            event_id=self._next_id(),
            timestamp_ns=self._now_ns(),
            event_type=EventType.ADD,
            order_id=order_id,
            side=side,
            price=price,
            qty=qty,
            best_bid=best_bid,
            best_ask=best_ask,
        )
        self.events.append(ev)
        return ev

    def log_cancel(self, order_id: int, best_bid: float | None, best_ask: float | None) -> Event:
        ev = Event(
            event_id=self._next_id(),
            timestamp_ns=self._now_ns(),
            event_type=EventType.CANCEL,
            order_id=order_id,
            best_bid=best_bid,
            best_ask=best_ask,
        )
        self.events.append(ev)
        return ev

    def log_modify(self, order_id: int, new_qty: int, best_bid: float | None, best_ask: float | None) -> Event:
        ev = Event(
            event_id=self._next_id(),
            timestamp_ns=self._now_ns(),
            event_type=EventType.MODIFY,
            order_id=order_id,
            qty=new_qty,
            best_bid=best_bid,
            best_ask=best_ask,
        )
        self.events.append(ev)
        return ev

    def log_fill(self, order_id: int, counterparty_id: int, trade_qty: int, price: float, side: str, filled_qty: int, best_bid: float | None, best_ask: float | None) -> Event:
        ev = Event(
            event_id=self._next_id(),
            timestamp_ns=self._now_ns(),
            event_type=EventType.FILL,
            order_id=order_id,
            side=side,
            price=price,
            counterparty_id=counterparty_id,
            trade_qty=trade_qty,
            filled_qty=filled_qty,
            best_bid=best_bid,
            best_ask=best_ask,
        )
        self.events.append(ev)
        return ev

    def to_dicts(self) -> list[dict[str, object]]:
        return [
            {
                "event_id": e.event_id,
                "timestamp_ns": e.timestamp_ns,
                "event_type": e.event_type.name,
                "order_id": e.order_id,
                "side": e.side,
                "price": e.price,
                "qty": e.qty,
                "filled_qty": e.filled_qty,
                "counterparty_id": e.counterparty_id,
                "trade_qty": e.trade_qty,
                "best_bid": e.best_bid,
                "best_ask": e.best_ask,
            }
            for e in self.events
        ]

    def __len__(self) -> int:
        return len(self.events)
